File: bindings/python/examples/grpc_stream_checkpoints.py
# Copyright (c) 2026 IOTA Stiftung
# SPDX-License-Identifier: Apache-2.0

from lib.iota_sdk import *

import asyncio


async def main():
    client = GrpcClient.new_testnet()

    # Pick a small range of recent checkpoints to stream.
    latest = await client.get_checkpoint_latest()
    start = latest.sequence_number - 4

    stream = await client.stream_checkpoints(start, latest.sequence_number)

    while (checkpoint := await stream.next()) is not None:
        summary = checkpoint.summary
        print(f"Checkpoint {checkpoint.sequence_number}:",
              f"epoch {summary.epoch()},",
              f"{summary.network_total_transactions()} total transactions,",
              f"timestamp {summary.timestamp_ms()}")


if __name__ == "__main__":
    asyncio.run(main())

File: crates/iota-sdk/examples/grpc_stream_checkpoints.rs
// Copyright (c) 2026 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

// TODO this has been added to develop already

//! Stream a range of checkpoints over gRPC.

use futures::StreamExt;
use iota_sdk::grpc_client::{Client, Result, read_mask_fields::CheckpointResponseReadMask};

#[tokio::main]
async fn main() -> Result<()> {
    let client = Client::new_testnet()?;

    // Pick a small range of recent checkpoints to stream.
    let latest = client
        .get_checkpoint_latest(None, None, CheckpointResponseReadMask::default())
        .await?
        .into_inner();
    let start = latest.sequence_number.saturating_sub(4);

    let mut stream = client
        .stream_checkpoints(
            Some(start),
            Some(latest.sequence_number),
            None,
            None,
            CheckpointResponseReadMask::default(),
        )
        .await?;

    while let Some(checkpoint) = stream.body_mut().next().await {
        let checkpoint = checkpoint?;
        let summary = checkpoint.summary()?.summary()?;
        println!(
            "Checkpoint {}: epoch {}, {} total transactions, timestamp {}",
            checkpoint.sequence_number,
            summary.epoch,
            summary.network_total_transactions,
            summary.timestamp_ms,
        );
    }

    Ok(())
}
